src/domain_bridge/utils.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright 2021, Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <limits>
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+#include "rcl/node.h"
+#include "rcl/node_options.h"
+#include "rcl/time.h"
+#include "rclcpp/context.hpp"
+#include "rclcpp/init_options.hpp"
+#include "rclcpp/node.hpp"
+#include "rclcpp/node_interfaces/node_base.hpp"
+
+#include "domain_bridge/utils.hpp"
+
+namespace domain_bridge
+{
+
+namespace utils
+{
+
+rclcpp::Duration from_rmw_time(rmw_time_t duration)
+{
+  constexpr rcl_duration_value_t limit_ns = std::numeric_limits<rcl_duration_value_t>::max();
+  constexpr rcl_duration_value_t limit_sec = RCL_NS_TO_S(limit_ns);
+  if (duration.sec > limit_sec || duration.nsec > limit_ns) {
+    return rclcpp::Duration{limit_ns};
+  }
+  uint64_t total_ns = RCL_S_TO_NS(duration.sec) + duration.nsec;
+  if (total_ns > limit_ns) {
+    return rclcpp::Duration{limit_ns};
+  }
+  return rclcpp::Duration{static_cast<rcl_duration_value_t>(total_ns)};
+}
+
+rclcpp::Node::SharedPtr
+create_node(
+  const std::string & name,
+  std::size_t domain_id,
+  std::shared_ptr<rclcpp::Context> context)
+{
+  if (context == nullptr) {
+    context = std::make_shared<rclcpp::Context>();
+  }
+  rclcpp::InitOptions init_options;
+  init_options.auto_initialize_logging(false);
+  context->init(0, nullptr, init_options);
+
+  rclcpp::NodeOptions node_options;
+  node_options.context(context)
+  .use_global_arguments(false)
+  .start_parameter_services(false)
+  .start_parameter_event_publisher(false);
+
+  auto node = std::make_shared<rclcpp::Node>(name, node_options);
+  auto node_base_interface = node->get_node_base_interface();
+  rcl_node_t * rcl_node_handle = node_base_interface->get_rcl_node_handle();
+  // Hacky work-around because setting domain ID is not a feature in the rclcpp layer
+  const_cast<rcl_node_options_t *>(
+    rcl_node_get_options(rcl_node_handle))->domain_id = domain_id;
```

Comment:
IIRC, since Foxy we no longer have a 1-to-1 mapping from ROS nodes to DDS participants. Instead, a participant is mapped to each ROS context. Therefore, the domain ID is tied to the ROS context, not the node. I suspect the problem is that the domain ID being set in the node options are not being applied.

Unfortunately, it doesn't look like an `rcl` API exists in Foxy to set the domain ID. I think what we want to do is set this member of the RMW struct `rmw_init_options`:

https://github.com/ros2/rmw/blob/6d23f9fc6bfe4616e1efeeefee7a927635a83afe/rmw/include/rmw/init_options.h#L52

I suppose one possible fix is to backport the missing `rcl` setter:

https://github.com/ros2/rcl/blame/master/rcl/include/rcl/init_options.h#L174

and set it here before creating the context (ie. backport https://github.com/ros2/rcl/pull/678).

---

Or, we could try setting the environment variable `ROS_DOMAIN_ID` before creating the ROS context. This is what eProsima's Integration Service does (or used to): https://github.com/eProsima/Integration-Service/blob/a76ccb63fbdb9f8193fd252c6c8d1277c5f1e731/packages/ros2/src/SystemHandle.cpp#L142-L149

Note, it also restores the value of `ROS_DOMAIN_ID` when it is done.

Reply:
I tried both setting/resetting the environment variable `ROS_DOMAIN_ID` and tried initializing a `rclcpp::Context` with a `rclcpp::InitOptions` constructed with a domain ID-set `rcl_init_options_t`, but neither seemed to work. Will have to investigate a little further. The rpr job will fail as expected though because https://github.com/ros2/rcl/pull/946 hasn't been released yet.